api/utils/aws/endpoint.go

```diff
@@ -0,0 +1,164 @@
+	"net"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// IsAWSEndpoint returns true if the input URI is an AWS endpoint.
+func IsAWSEndpoint(uri string) bool {
+	// Note that AWSCNEndpointSuffix contains AWSEndpointSuffix so there is no
+	// need to search for AWSCNEndpointSuffix explicitly.
+	return strings.Contains(uri, AWSEndpointSuffix)
+}
+
+// IsRDSEndpoint returns true if the input URI is an RDS endpoint.
+//
+// https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/Aurora.Overview.Endpoints.html
+func IsRDSEndpoint(uri string) bool {
+	return strings.Contains(uri, RDSEndpointSubdomain) &&
+		IsAWSEndpoint(uri)
+}
+
+// IsRedshiftEndpoint returns true if the input URI is an Redshift endpoint.
+//
+// https://docs.aws.amazon.com/redshift/latest/mgmt/connecting-from-psql.html
+func IsRedshiftEndpoint(uri string) bool {
+	return strings.Contains(uri, RedshiftEndpointSubdomain) &&
+		IsAWSEndpoint(uri)
+}
+
+// trimAWSParentDomain removes common AWS endpoint suffixes from the endpoint.
+func trimAWSParentDomain(endpoint string) (string, error) {
+	if strings.HasSuffix(endpoint, AWSEndpointSuffix) {
+		return endpoint[:len(endpoint)-len(AWSEndpointSuffix)], nil
+	}
+
+	if strings.HasSuffix(endpoint, AWSCNEndpointSuffix) {
+		return endpoint[:len(endpoint)-len(AWSCNEndpointSuffix)], nil
+	}
+
+	return endpoint, trace.BadParameter("endpoint %v is not an AWS endpoint", endpoint)
+}
+
+// ParseRDSURI extracts the identifier and region from the provided RDS URI.
+func ParseRDSURI(uri string) (id, region string, err error) {
+	endpoint, _, err := net.SplitHostPort(uri)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+	return ParseRDSEndpoint(endpoint)
+}
+
+// ParseRDSEndpoint extracts the identifier and region from the provided RDS
+// endpoint.
+//
+// RDS/Aurora endpoints look like this:
+// aurora-instance-1.abcdefghijklmnop.us-west-1.rds.amazonaws.com
+// aurora-instance-2.abcdefghijklmnop.rds.cn-north-1.amazonaws.com.cn
+func ParseRDSEndpoint(endpoint string) (id, region string, err error) {
+	trimmedEndpoint, err := trimAWSParentDomain(endpoint)
+	if err != nil {
+		return "", "", trace.Wrap(err)
+	}
+
+	parts := strings.Split(trimmedEndpoint, ".")
+	if len(parts) != 4 {
+		return "", "", trace.BadParameter("failed to parse %v as RDS endpoint", endpoint)
+	}
+
+	// Service name/region can be at either position 2 or 3.
+	switch {
+	case parts[3] == RDSServiceName:
+		region = parts[2]
+
+	case parts[2] == RDSServiceName:
+		region = parts[3]
+
+	default:
+		return "", "", trace.BadParameter("failed to parse %v as RDS endpoint", endpoint)
+	}
+
+	return parts[0], region, nil
+}
+
+// ParseRedshiftURI extracts cluster ID and region from the provided Redshift
+// URI.
+//
+// Redshift endpoints look like this:
+// redshift-cluster-1.abcdefghijklmnop.us-east-1.rds.amazonaws.com
+// redshift-cluster-2.abcdefghijklmnop.rds.cn-north-1.amazonaws.com.cn
+func ParseRedshiftURI(uri string) (clusterID, region string, err error) {
```

Comment:
I'm not a huge fan of regex, but in this case I think that something like `.*\.([\w\d-]+.rds|rds.[\w\d-]+).amazonaws.com.*` looks cleaner.

Reply:
i don't like regex either in general either but it may not be a bad idea in this case. let me try it out. there are other partitions we don't support yet that actually has suffix like `.sc2s.sgov.gov` but those can be added to regex too.

Reply:
```
var redshiftRegex = regexp.MustCompile(`^(?P<id>[\w\d-]+)\.(?:[\w\d-]+)\.(?:redshift\.(?P<region>[\w\d-]+)|(?P<region>[\w\d-]+)\.redshift)\.amazonaws.com(?:\.cn)?$`)
                                                                                                                                  
func ParseRedshiftURI(uri string) (clusterID, region string, err error) {                                                        
    endpoint, _, err := net.SplitHostPort(uri)                                                                                    
    if err != nil {                                                                                                               
        return "", "", trace.Wrap(err)                                                                                            
    }                                                                                                                             
                                                                                                                                  
    values := redshiftRegex.FindStringSubmatch(endpoint)                                                                          
    if len(values) != 4 {                                                                                                         
        return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", uri)                                         
    }                                                                                                                             
                                                                                                                                  
    // values[0] is the capture of the entire string.
    // values[1] is <id>.
    // values[2] and values[3] is <region> depends on the position.
    return values[1], (values[2] + values[3]), nil

}
```

This is the best i can do with my limited regex knowledge. The regex gets complicated with capture groups. not sure if this reads better than current version.

WDYT?

Reply:
In my opinion this is cleaner than the current implementation with many `split` and boundy checks. I'd just add a comment with all strings that should match that regex or even a UT to make it clear what we want to match.

@r0mant @smallinsky Any opinions?

Reply:
```go
       parts := strings.Split(trimmedEndpoint, ".")
	if len(parts) != 4 {
		return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", endpoint)
	}

	// Service name/region can be at either position 2 or 3.
	switch {
	case parts[3] == RedshiftServiceName:
		region = parts[2]

	case parts[2] == RedshiftServiceName:
		region = parts[3]

	default:
		return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", endpoint)
	}

	return parts[0], region, nil
```
vs 
```go
    var redshiftRegex = regexp.MustCompile(`^(?P<id>[\w\d-]+)\.(?:[\w\d-]+)\.(?:redshift\.(?P<region>[\w\d-]+)|(?P<region>[\w\d-]+)\.redshift)\.amazonaws.com(?:\.cn)?$`)
 ...
    values := redshiftRegex.FindStringSubmatch(endpoint)                                                                          
    if len(values) != 4 {                                                                                                         
        return "", "", trace.BadParameter("failed to parse %v as Redshift endpoint", uri)                                         
    }                                                                                                                             
                                                                                                                                  
    // values[0] is the capture of the entire string.
    // values[1] is <id>.
    // values[2] and values[3] is <region> depends on the position.
    return values[1], (values[2] + values[3]), nil
```

Personally I would incline to the manual parsing version. Regex for me is like blackbox and really hard to understand only by reading but this is only my preference.

Reply:
I think that manual parsing is easier to understand also, but I would consider splitting this into `parseRedshiftURI` and `parseRedshiftCNURI` similar to my RDS suggestion above which would make it a little more cleaner (I think).

Reply:
I have an upcoming change for RDS proxy which will make this parsing even messier. And another potential change is the redshift serverless

How about something like this
```
var rdsEndpointFormats = []string{                                               
    "<id>.<hash>.<region>.rds.amazonaws.com",                                       
    "<id>.<hash>.rds.<region>.amazonaws.com.cn",                                    
}                                
                                                                                                       
func ParseRDSEndpoint(endpoint string) (id, region string, err error) {                                
    for _, format := range rdsEndpointFormats {                                
        if values, ok := parseEndpointWithFormat(endpoint, format); ok {                       
            return values["id"], values["region"], nil                                
        }                                             
    }                                                 
    return "", "", trace.BadParameter("failed to parse %v as RDS endpoint", endpoint)                                
}  
```
where format is a custom format that has keys surrounded by `<>`

I have implemented `parseEndpointWithFormat` locally, which a small function (about 30lines) that translate this format to a regex for matching and capturing the values.

Obviously performance won't be as good as manual parsing, but it's easier to add more formats later.

What do you guys think? @r0mant @smallinsky @jakule